primary/src/certificate_waiter.rs

```diff
@@ -160,12 +160,20 @@ impl<PublicKey: VerifyingKey> CertificateWaiter<PublicKey> {
             let round = self.consensus_round.load(Ordering::Relaxed);
             if round > self.gc_depth {
                 let gc_round = round - self.gc_depth;
-                for (r, cancel_handle) in self.pending.values() {
-                    if *r <= gc_round + 1 {
-                        let _ = cancel_handle.send(()).await;
-                    }
-                }
-                self.pending.retain(|_, (r, _)| *r > gc_round + 1);
+                self.pending = self
+                    .pending
+                    .drain()
+                    .flat_map(|(digest, (r, handler))| {
+                        if r <= gc_round {
```

Comment:
Note how `*r <= gc_round + 1` becomes `r <= gc_round`, leaving more certs in store here.